torch/_higher_order_ops/associative_scan.py

```diff
@@ -79,32 +109,150 @@ def add(x: torch.Tensor, y: torch.Tensor):
 
     if not torch._dynamo.is_compiling():
         with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
-            return torch.compile(associative_scan, fullgraph=True)(
-                combine_fn, input, dim
-            )
+            if host_side:
+                return torch.compile(associative_scan_host_side, fullgraph=True)(
+                    combine_fn, input, dim, reverse
+                )
+                # return associative_scan_host_side(combine_fn, input, dim, False)
+            else:
+                if reverse:
+                    raise ValueError(
+                        "For the device-side associative scan, the reverse flag has not yet been implemented"
+                    )
+
+                return torch.compile(associative_scan, fullgraph=True)(
+                    combine_fn, input, dim
+                )
 
     leaves, spec = pytree.tree_flatten(input)
 
-    assert len(leaves) >= 1, "expected at least 1 input leaf"
-    assert all(
-        isinstance(x, torch.Tensor) for x in leaves
-    ), "input leaves must be a Tensor"
-    shape = leaves[0].shape
-    ndim = len(shape)
-    dim = utils.canonicalize_dim(ndim, dim)
+    check_args(input, combine_fn, leaves, spec, dim)
 
-    for x in leaves[1:]:
-        assert x.shape == shape, "All input tensors must have the same shape"
+    if reverse and host_side:
+        raise ValueError(
+            "For the device-side associative scan, the reverse flag has not yet been implemented"
+        )
 
     combine_fn = functools.partial(
         wrap_combine_fn_flat, combine_fn=combine_fn, spec=spec, num_leaves=len(leaves)
     )
 
-    result_flat = associative_scan_op(combine_fn, leaves, dim)
+    if host_side:
+        with _set_compilation_env():
+            with torch._dynamo.utils.disable_cache_limit():
+                with _temp_remove_pre_dispatch_torch_function_mode():
+                    return torch.compile(
+                        associative_scan_host_side, backend="eager", fullgraph=True
+                    )(combine_fn, input, dim, reverse)
+    else:
+        result_flat = associative_scan_op(combine_fn, leaves, dim)
 
     return pytree.tree_unflatten(result_flat, spec)
 
 
+def _interleave(a, b, dim):
+    # https://stackoverflow.com/questions/60869537/how-can-i-interleave-5-pytorch-tensors
+    if b_trunc := (a.shape[dim] == b.shape[dim] + 1):
+        pad = [0, 0] * b.ndim
+        pad[
+            (b.ndim - dim - 1) * 2 + 1
+        ] = 1  # +1=always end of dim, pad-order is reversed so start is at end
+        b = torch.nn.functional.pad(b, pad)
+
+    stacked = torch.stack([a, b], dim=dim + 1)
+    interleaved = torch.flatten(stacked, start_dim=dim, end_dim=dim + 1)
+    if b_trunc:
+        # TODO: find torch alternative for slice_along dim for torch.jit.script to work
```

Comment:
Regarding `slice_along_dim`, maybe this feature request is related: https://github.com/pytorch/pytorch/issues/41625

Reply:
Thanks for the pointer! I tried to follow the issues, but is there anything at the moment that I should incorporate?

Reply:
I'm not a maintainer/reviewer. I also missed a more capable/simple slicing function to not have to juggle slice(None) and calculate the correct dimension index, that's all :)